whylog/constraints/__init__.py

```diff
@@ -1,52 +1,202 @@
+
+    def _check_useless_params(self, correct_param_names, actual_param_names):
+        if actual_param_names - correct_param_names:
+            raise ConstructorParamsError(self.TYPE, correct_param_names, actual_param_names)
+
+    def _check_mandatory_params(self, correct_param_names, actual_param_names):
+        pass
 
-class IdenticalConstraint(AbstractConstraint):
-    def __init__(self, groups, params):
+    def _check_optional_params(self, correct_param_names, actual_param_names):
         pass
 
+    def convert_to_user_constraint_intent(self):
+        """
+        Converts constraint to UserConstraintIntent object.
+
+        For Teacher and Config use while saving constraint into Whylog knowledge base.
+        """
+        return UserConstraintIntent(self.TYPE, self.groups, self.params)
+
     @classmethod
-    def verify(cls, group_contents, param_dict):
-        if len(group_contents) <= 1:
-            return False  # FIXME raise exception?
-        return all(group_contents[0] == group for group in group_contents)
+    def get_param_names(cls):
+        """
+        Returns names of constraint additional params.
 
+        For Front to display param names to user and then ask user for param contents.
+        """
+        return cls.PARAMS
 
-class AnyValueIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
+    @classmethod
+    def get_groups_count(cls):
+        """
+        Returns minimal and maximal count of groups needed to create constraint.
 
+        For Front to ask user for proper count of groups.
+        2, None - at least 2 groups
+        2, 2 - exactly 2 groups
+        """
+        return cls.MIN_GROUPS_COUNT, cls.MAX_GROUPS_COUNT
 
-class DifferentValueIntervals(AbstractConstraint):
-    def __init__(self, intervals):
-        pass
+    @abstractmethod
+    def verify(self, group_contents, param_dict):
+        """
+        Verifies constraint for given params in param_dict and groups contents.
+
+        :param param_dict: dict of additional params of constraint
+        :param groups: list of groups contents,
 
+        For LogReader and Teacher verification.
+        It must be optimized as well as possible (for LogReader).
+        """
 
-class ConstIntervals(AbstractConstraint):
-    def __init__(self, intervals):
         pass
 
 
-class ValueDeltaIntervals(AbstractConstraint):
-    def __init__(self, interval_lower, interval_greater, min_delta=None, max_delta=None):
+class TimeConstraint(AbstractConstraint):
+    """
+    Time delta between two dates must be greater or equal to 'min_delta'
+    and lower or equal to 'max_delta'
+    """
+
+    TYPE = ConstraintType.TIME_DELTA
+
+    MIN_GROUPS_COUNT = 2
+    MAX_GROUPS_COUNT = 2
+
+    MIN_DELTA = 'min_delta'
+    MAX_DELTA = 'max_delta'
+
+    PARAMS = sorted([MIN_DELTA, MAX_DELTA])
+
+    def __init__(self, groups=None, param_dict=None, params_checking=True):
```

Comment:
constructor from abstract class can be used here. Please move the docstring to the class level and remove this constructor.